Improve bound checking laromance

sapitts opened the pull request:
Includes two commits to allow easy adaption based on discussion:
  1. Corrects the logic operator and checks the ROM output values in `computeStressFinalize`
  2. Replaces the use of the `checkInputWindows` method in `computeStressFinalize` to throw an exception and cut the timestep if the calculated dislocation densities are outside of the allowable evolution bounds
    - This approach leverages the treatment in the ROM to reset the input effective strain value at the maximum allowable, regardless of the actual simulation value (per email conversations)

fyi @lynnmunday @tophmatthews @bwspenc 

Marked as WIP only to prevent testing while Civet is down

Refs #14046

sapitts committed:
Improve error checking on allowable bounds, Refs #14046

tophmatthews commented on modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:
```diff
@@ -438,7 +438,7 @@ ADLAROMANCEStressUpdateBase::checkInputWindow(const ADReal & input,
                                               const std::vector<Real> & global_limits)
 {
-  if (behavior != WindowFailure::WARN || behavior != WindowFailure::ERROR)
+  if (behavior != WindowFailure::WARN && behavior != WindowFailure::ERROR)
```
Whoops!

tophmatthews commented on modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:
```diff
@@ -794,8 +794,29 @@ ADLAROMANCEStressUpdateBase::computeStressFinalize(const ADRankTwoTensor & plast
   _wall_dislocations[_qp] = _old_input_values[_wall_output_index] + _wall_dislocation_increment;
 
-  for (unsigned int i = 0; i < _num_inputs; i++)
-    checkInputWindow(_input_values[i], _window_failure[i], _global_limits[i]);
```
This skips the other input values no? Do the checking with the error correction above not work?

sapitts replied:
In my print statements I was seeing that this line was checking the values input into the ROM while I would like to have a check for the output ROM values: this line would check the dislocations calculated at timestep (n-1) and I'd like to have a check on the dislocations calculated at timestep (n)

tophmatthews replied:
Oh I see. I think you can just pass in output values, eg.

```
for (unsigned int i = 0; i < _num_outputs; i++)
    checkInputWindow(_output_values[i], _window_failure[i], _global_limits[i]);
```

sapitts replied:
Yes, I agree that option does work; I did this in the first commit so that we can drop the second commit if the exception throwing isn't desired

tophmatthews replied:
The main problem here is that `_global_limits` is really only applicable to the starting values of the input, and not the only value of the input. I had a similar check in there before the latest changes, but removed them based on suggestions from Aaron that there is no real lower limit. If you want to put `0.0` as the lower limit, that is probably fine:

```
for (unsigned int i = 0; i < _num_outputs; i++)
    checkInputWindow(_output_values[i], _window_failure[i], 0.0);
```

tophmatthews replied:
> no real lower limit

Rather, the global limits are not the actual limits of the ROM.

sapitts replied:
Are you suggesting that we remove the check on the ROM outputs at the end of the timestep all together then?

sapitts replied:
This class is the only one that evolves the dislocation densities, so isn't it better do check if the calculated dislocations values from this timestep will be allowable as inputs in the next while we can still do something (e.g. cut the timestep) about it and try to get the simulation to run a bit further?

tophmatthews replied:
> Are you suggesting that we remove the check on the ROM outputs at the end of the timestep all together then?

Yes, the confusing bit is that the rom (should) only develop dislocation densities that make sense. The parameters here are only valid for the initial dislocation densities, from which to evolve from, not what the absolute lowest limit is. We should try and have a call with Aaron and Laurent to clarify.

tophmatthews replied:
>This class is the only one that evolves the dislocation densities, so isn't it better to check if the calculated dislocations values from this timestep will be allowable as inputs in the next while we can still do something (e.g. cut the timestep) about it and try to get the simulation to run a bit further?

There is no "valid lower limit" of densities determined by the ROM. I would agree that if they are unphysical (i.e. negative) then it should cut the timestep, hence my latest code snippet that should be used.

tophmatthews replied:
Let's confirm, but I think the check you are adding here, and the check I have previously and you removed, are both unnecessary.

sapitts replied:
I had thought, having been corrected previously last week by Aaron, that the initial condition values were the values we set with the `params.addRangeCheckedParam` and the evolution values were set by the `_global_limits`. Clarification on this from Aaron and Laurent would be useful

sapitts replied:
> hence my latest code snippet that should be used
> ```
> for (unsigned int i = 0; i < _num_outputs; i++)
>    checkInputWindow(_output_values[i], _window_failure[i], 0.0);
>```
This won't work as written because there isn't a correspondence between the `_output_values` vector and the `_window_failure` vector. I have something similar using the index variables you created waiting in the wings. No sense in pushing it just to delete it

tophmatthews replied:
Ah yes...well it would just be a hardcoding of the three output indexes then:
```
  checkInputWindow(_output_values[_cell_output_index], _window_failure[_cell_input_index], 0.0);
  checkInputWindow(_output_values[_wall_output_index], _window_failure[_wall_input_index], 0.0);
  checkInputWindow(_output_values[_strain_output_index], _window_failure[_stress_input_index], 0.0);
```

tophmatthews replied:
Well...not totally either...but you get the idea. Basically, pass the values to `checkInputWindow` so you don't have to duplicate so much code.

tophmatthews replied:
You don't need fuzzy on 0.0 (I think). Why not use `checkInputWindow`?

sapitts replied:
>Why not use `checkInputWindow`?
Doesn't compile: it needs a vector not a constant.  I'm happy to continue discussing the best way to give the most useful information with the least amount of code once we hear from Laurent and Aaron, but if they say this check isn't necessary then this discussion is all wasted effort

tophmatthews replied:
Okay, sounds like it was what I was thinking. I still think you are correct to have a sanity check on physical values for the dislocations here. I wonder if you'd just want to throw the exception, like you had previously, if one of the output values is unphysical:
```
  if (_cell_dislocations[_qp] < 0.0 || _wall_dislocations[_qp] < 0.0)
  {
    mooseExepction("In ",
                   _name,
                   ": Negative disclocation density calculated for (",
                   _cell_dislocations[_qp],
                   ") or wall (",
                   _wall_dislocations[_qp],
                   ").");
    _cell_dislocations[_qp] = _old_input_values[_cell_output_index];
    _wall_dislocations[_qp] = _old_input_values[_wall_output_index];
  }
```

Note the last bit of setting the dislocations to their old value is to prevent an unphysical value being passed else where and breaking another piece of code. I have seen this before with things like thermal conductivity that resulted in feedback loop that prevented a proper recovery via an exception. Yes, the dislocation densities are only used here, but it's good practice to protect future problems by allowing a graceful failure.

sapitts replied:
I like it. If Aaron gives us the okay I'll add this in and repush

sapitts committed:
Address review comments, Refs #14046

sapitts committed:
Address reviewer comments round 2:

Use of a virtual function to set the rom strain output
cutoff value in inherited classes

Refs #14046

bwspenc commented:
go away, stale bot

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/586028/) on 237ee9c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15746/site)

This comment will be updated on new commits.

sapitts commented:
@loganharbour would you or one of the other MOOSE folks give me some help with identifying what's wrong with the private app PETSc alt tests please? I don't have the permissions to see the testing details and am quite in the dark about what the error might be

moosebuild commented:
Job [Private App tests PETSc alt](https://civet.inl.gov/job/586033/) on 237ee9c : invalidated by @loganharbour

Unrelated failure

loganharbour commented:
> what's wrong with the private app PETSc alt tests please?

Everything failing/allowed to fail is unrelated! I'm gonna put in a change in civet shortly to avoid the failure you just hit - it should be allowed to fail as well.

sapitts commented:
> Everything failing/allowed to fail is unrelated!

This is my most favorite thing to hear about a testing failure on one of my PRs :-) Thanks for invalidating and adding the change on civet

loganharbour commented:
👍🏼 sure thing! This one might fail again while the recipe change propagates, in which case it'll need one more invalidation. I'll keep an eye on it.

sapitts commented:
@dschwen would you review this PR while Ben is out? It will help in your LAROMANCE simulations to some extent

tophmatthews commented on modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:
```diff
@@ -89,6 +89,10 @@ ADLAROMANCEStressUpdateBase::validParams()
       "Optional forcing function for the creep strain from the previous timestep. If provided, "
       "the old creep strain will be reset to the function value at the beginning of the "
       "timestep. Used for testing purposes only.");
+  params.addParam<Real>("rom_strain_cutoff",
```
This shouldn't be a input parameter, since it is ROM specific, and won't change. I'd consider putting it directly into the inherited classes as another material specific ROM value.

tophmatthews replied:
Won't change in a given simulation I mean...it'll change between ROM's, rather it'll never change between different simulations using the same ROM...

dschwen replied:
`virtual Real romStrainCutoff() = 0;` ?

sapitts replied:
virtual function added, and override in the inheriting test material added

tophmatthews reviewed (changes requested):
I'm trying to remember what else we had going here, but I think this mostly captured the things we talked about a couple months ago, thanks @sapitts! Just the one fix for the cutoff.

dschwen commented on modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:
```diff
@@ -692,10 +697,11 @@ ADLAROMANCEStressUpdateBase::convertOutput(const std::vector<Real> & old_input_v
   ADReal expout = std::exp(rom_output);
-  if (expout > 1.0e-10)
-    expout -= 1.000000E-10;
+  if (expout > _rom_strain_output_cutoff)
+    expout -= _rom_strain_output_cutoff;
```
I don't understand whats going on here. Let's say `expout` is 4*`_rom_strain_output_cutoff`. Then you subtract it once, but it is still larger than the cutoff value... ?

tophmatthews replied:
Yeah...I'm not sure why it's like this either. This is a part of the way LAROMANCE does things and how the ROMs are formulated...

sapitts replied:
Is there a change here that's desired? A general comment string that we're only checking for low bounds? Or is okay as is?

dschwen commented on modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:
```diff
@@ -692,10 +697,11 @@ ADLAROMANCEStressUpdateBase::convertOutput(const std::vector<Real> & old_input_v
     return 0.0;
 
   ADReal expout = std::exp(rom_output);
```
This is not never 0.0! E.g.`std::exp(-800.0) == 0.0`

tophmatthews replied:
? rom_output isn't necessarily zero or negative...

dschwen replied:
We divide by expout below without further checks. I suggest adding an assertion to make sure expout > 0

dschwen replied:
Oh, I see rom_output is guaranteed to be positive?

tophmatthews replied:
Uh....I'm not sure it's guaranteed. Maybe. Certainly an assert is easy to add

sapitts replied:
Assert added

dschwen reviewed (commented):
See individual comments

dschwen commented:
I'm currently on a several day journey to hunt down the source of an Inf in
NEML. And let me tell you, a few assertions could save me a lot of time...

On Thu, Oct 15, 2020, 7:53 AM Topher Matthews <notifications@github.com>
wrote:

> *@tophmatthews* commented on this pull request.
> ------------------------------
>
> In modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C
> <https://github.com/idaholab/moose/pull/15746#discussion_r505560488>:
>
> > @@ -692,10 +697,11 @@ ADLAROMANCEStressUpdateBase::convertOutput(const std::vector<Real> & old_input_v
>      return 0.0;
>
>    ADReal expout = std::exp(rom_output);
>
> Uh....I'm not sure it's guaranteed. Maybe. Certainly an assert is easy to
> add
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/15746#discussion_r505560488>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABRMPSYGC5QAVCTTZ3F43TSK35FDANCNFSM4P2M7JJQ>
> .
>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/605672/) on 8447144 : invalidated by @milljm

updated ssh keys

sapitts commented:
@bwspenc and/or @dschwen I believe I've address (or asked a question about) all of the comments here. Would one of you take another look at this please? We need this change in so that we can get Topher's merge request in Blackbear in

moosebuild commented:
Job [Mac Test](https://civet.inl.gov/job/623200/) on 2645bd7 : invalidated by @sapitts

Unrelated porous flow test failure

dschwen commented on modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:
```diff
@@ -692,10 +692,13 @@ ADLAROMANCEStressUpdateBase::convertOutput(const std::vector<Real> & old_input_v
   ADReal expout = std::exp(rom_output);
-  if (expout > 1.0e-10)
-    expout -= 1.000000E-10;
+  mooseAssert(expout > 0.0, "ROM calculated strain increment is zero or negative");
```
How is this assertion ever hit?! The return value of `std::exp` should always be > 0... unless `rom_output` is `NaN` or (maybe) `-inf`.

dschwen replied:
It can certainly never be negative

sapitts replied:
> We divide by expout below without further checks. I suggest adding an assertion to make sure expout > 0

Did you want the assert from the Oct 14 review above the exponential call? I'm happy to move it and I'm just not sure where you'd like to see the assert

dschwen replied:
Ah, touche :-). How about changing the assertion text to "is not strictly positive"?

sapitts replied:
done

dschwen commented on modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C:
```diff
@@ -692,10 +692,13 @@ ADLAROMANCEStressUpdateBase::convertOutput(const std::vector<Real> & old_input_v
+  mooseAssert(expout > 0.0, "ROM calculated strain increment is zero or negative");
+
+  const Real rom_strain_cutoff_value = romStrainCutoff();
+  if (expout > rom_strain_cutoff_value)
+    expout -= rom_strain_cutoff_value;
```
I do not understand this conditional. It introduces a weird discontinuity.

dschwen replied:
My expectation would have been `expout = rom_strain_cutoff_value;`

dschwen replied:
no it does not

sapitts committed:
Refined addressing of reviewer requirements, Refs #14046
